src/uct/base/uct_component.h

```diff
@@ -14,33 +14,127 @@
+ */
+typedef ucs_status_t (*uct_component_rkey_unpack_func_t)(
+                uct_component_t *component, const void *rkey_buffer,
+                uct_rkey_t *rkey_p, void **handle_p);
+
+
+/**
+ * Component method to obtain a locally accessible pointer to a remote key.
+ *
+ * @param [in]  component               Get remote key memory pointer on this
+ *                                      component.
+ * @param [in]  rkey                    Obtain the pointer for this remote key.
+ * @param [in]  handle                  Remote key handle, as returned from
+ *                                      @ref uct_component_rkey_unpack_func_t.
+ * @param [in]  remote_addr             Remote address to obtain the pointer for.
+ * @param [out] local_addr_p            Filled with the local access pointer.
+ *
+ * @return UCS_OK on success or error code in case of failure.
+ */
+typedef ucs_status_t (*uct_component_rkey_ptr_func_t)(
+                uct_component_t *component, uct_rkey_t rkey, void *handle,
+                uint64_t remote_addr, void **local_addr_p);
 
-    ucs_status_t           (*rkey_unpack)(uct_md_component_t *mdc, const void *rkey_buffer,
-                                          uct_rkey_t *rkey_p, void **handle_p);
 
-    ucs_status_t           (*rkey_ptr)(uct_md_component_t *mdc, uct_rkey_t rkey, void *handle,
-                                       uint64_t raddr, void **laddr_p);
+/**
+ * Component method to release an unpacked remote key.
+ *
+ * @param [in]  component               Release the remote key of this
+ *                                      component.
+ * @param [in]  rkey                    Release this remote key.
+ * @param [in]  handle                  Remote key handle, as returned from
+ *                                      @ref uct_component_rkey_unpack_func_t.
+ *
+ * @return UCS_OK on success or error code in case of failure.
+ */
+typedef ucs_status_t (*uct_component_rkey_release_func_t)(
+                uct_component_t *component, uct_rkey_t rkey, void *handle);
 
-    ucs_status_t           (*rkey_release)(uct_md_component_t *mdc, uct_rkey_t rkey,
-                                           void *handle);
 
-    const char             name[UCT_MD_COMPONENT_NAME_MAX];
-    void                   *priv;
-    ucs_config_global_list_entry_t md_config;  /**< MD configuration entry */
-    ucs_list_link_t        tl_list;            /**< List of uct_md_registered_tl_t */
-    ucs_list_link_t        list;
+/**
+ * Defines a UCT component
+ */
+struct uct_md_component {
+    const char                              name[UCT_MD_COMPONENT_NAME_MAX]; /**< Component name */
+    uct_component_query_md_resources_func_t query_md_resources; /**< Query memory domain resources method */
+    uct_component_md_open_func_t            md_open;            /**< Memory domain open method */
+    uct_component_cm_open_func_t            cm_open;            /**< Connection manager open method */
```

Comment:
why cm_open is a method of "md_component"? shouldn't it be a part of "component"?

Reply:
i will rename uct_md_component  to uct_component in next PR (was split because of size)

Reply:
Ok, just to be aware, will it be a single structure or some inheritance?

Reply:
same single structure
just renaming it

Reply:
thanks, as discussed offline it will be a single structure.